### DISCUSSION TITLE
Weird behavior when Textual App started under multiprocessing.Process

### DISCUSSION THREAD

**Discussion by frankie567** [Textualize/textual#601] [Q&A] [answered]
Hello there 👋 

I was experimenting to implement an app which would have two processes: a "background" one that would actually run the heavy logic of the app, and a "ui" one which would run the Textual app. Both would communicate through an IPC socket.

So I started with a simple Textual app and tried to run it with `multiprocessing.Process`:

```py
from multiprocessing import Process
from textual.app import App
from textual.widgets import Placeholder


class SimpleApp(App):

    async def on_mount(self) -> None:
        await self.view.dock(Placeholder(), edge="left", size=40)
        await self.view.dock(Placeholder(), Placeholder(), edge="top")


def run():
    SimpleApp.run(log="textual.log")


if __name__ == '__main__':
    p = Process(target=run)
    p.start()
```

The app starts well, but whenever I move the mouse, click somewhere, escape sequences (or whatever there are 🙃) are displayed under the UI:

<img width="1837" alt="Capture d’écran 2022-07-03 à 11 00 58" src="https://user-images.githubusercontent.com/1144727/177032760-b0d899e7-472d-471f-8c5a-62334e76cd4e.png">

Textual App doesn't seem able to react to those events. My guess is that the console events are catched by the parent process rather than the Textual process, which cause this behavior. Is there something to do to circumvent this?

Cheers!

**Comment by frankie567** [comment, marked as answer]
After digging, I found the source of the problem and the issue. On macOS and Windows, when using `multiprocess.Process`, Python will spawn a new process, [which creates new file descriptors, in particular for `stdin`](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods).

Then, the `LinuxDriver` will be initialized inside the sub-process with this new `stdin`, different from the main process:

https://github.com/Textualize/textual/blob/e406458b8917d660197b8eeafcc1b7fde01e32a6/src/textual/drivers/linux_driver.py#L27-L32

Therefore, events won't be catched by the Textual app since they will be forwarded to the main `stdin` which started the process. I'm not sure why the escape sequences are displayed in the console in this case; maybe it's because the driver sets some options on the console.

Anyway, the solution I found is to subclass the driver so it accepts an arbitrary file descriptor. This way, I can forward the main `stdin` when starting the process. And it seems to work!

```py
import functools
import sys
from multiprocessing import Process

from rich.console import Console
from textual._types import MessageTarget
from textual.app import App
from textual.drivers.linux_driver import LinuxDriver
from textual.widgets import Placeholder


class SimpleApp(App):

    async def on_mount(self) -> None:
        await self.view.dock(Placeholder(), edge="left", size=40)
        await self.view.dock(Placeholder(), Placeholder(), edge="top")


class LinuxStdInDriver(LinuxDriver):
    def __init__(self, stdin_fd: int, console: Console, target: MessageTarget) -> None:
        super().__init__(console, target)
        self.fileno = stdin_fd


def run(stdin_fileno: int):
    driver_class = functools.partial(LinuxStdInDriver, stdin_fileno)
    SimpleApp.run(driver=driver_class, log="textual.log")


if __name__ == '__main__':
    stdin_fileno = sys.stdin.fileno()
    p = Process(target=run, args=(stdin_fileno,))
    p.start()
```

The solution is quite clunky with the current API. Maybe it could be worth to be able to manually set the `stdin` and `stdout` file descriptors when running the app? The `Console` class can already do something similar with the `file` argument to set the output file.

**Comment by Erotemic** [comment]
I'm seeing something I think might be similar, although I'm on Linux. I'm not 100% sure it's the same or if its even caused by textual, but I _think_ it is, and I at least want to document the issue even if I migrate it somewhere else later.

What I'm seeing is that when I am in an ssh session, and I start tmux, run a textual app, and then detatch from tmux strange sequences of characters are printted across the screen. 

![image](https://user-images.githubusercontent.com/3186211/181558727-2f7581a0-ef4f-4221-b636-195dc22c7268.png)

I can't select anything or interact with the terminal, except for ctrl C or enter which tries to enter the command and it fails. But then it keeps on typing as long as I have focus on the terminal with the ssh session in it.

![image](https://user-images.githubusercontent.com/3186211/181561577-60e6432a-f217-4052-82ae-360aabc7b9a0.png)

... Apparently, if I type really fast, I can get a command in. But then the random characters start again, but only if my mouse moves... very strange. I'm thinking textual doing something with stdin in combo with tmux might be the issue, but idk. 

Not sure how many of these variables can be removed to make a MWE, haven't gotten to that yet. Will post more details as I have them.
